pkg/build/api/validation/validation.go

```diff
@@ -145,23 +147,63 @@ func validateSource(input *buildapi.BuildSource) fielderrors.ValidationErrorList
 			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
 		}
 		if input.Dockerfile != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("dockerfile", "", "may not be set when type is Git"))
+			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+		}
+		if input.Binary != nil {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
+		}
+	case buildapi.BuildSourceBinary:
+		if input.Binary == nil {
+			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
+		} else {
+			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
+		}
+		if input.Dockerfile != nil {
+			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+		}
+		if input.Git != nil {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))
```

Comment:
>  Layering requires order decisions, and that scares me.

except that you did sign up for layering with the separate Dockerfile + binarysource support.  but yes in general i agree.